Pass tap by fd

bryteise opened the pull request:
Make use of the cloud-hypervisor change to support tap fds passed in being able to stand being rebooted. Revert the change to passing the tap name and add a fix to account for the queue in libvirt's domain conf means queue pairs (num_queues should be 2x queue pairs).

bryteise committed:
Revert "ch: pass TAP device name to CLH instead of fds"

This reverts commit 3a246c5e91fc8defa35ee72ffd69b33d1e9da5cf.

Use of fds instead of device names is now supported fully in
cloud-hypervisor so go back to that approach.

bryteise committed:
Don't set persistence for transient domains

bryteise committed:
Only set shutdown status for persistent domains

bryteise committed:
ch: Correctly set queue pairs for tapfdsize

bryteise commented:
@praveen-pk @russell-islam FYI

russell-islam commented:
LGTM

praveen-pk commented:
I was about to push a PR as well, you beat me to it :)

praveen-pk commented on src/ch/ch_process.c:
```diff
@@ -608,20 +607,14 @@ chProcessNetworkPrepareDevices(virCHDriverPtr driver, virDomainObjPtr vm)
     cleanup:
         g_free(tapfd);
-    return -1;
+    return 0;
```
Can you change this to "return -1"? In case of a failure this should return -1 here.

bryteise replied:
Done.

praveen-pk commented on src/ch/ch_monitor.c:
```diff
@@ -406,7 +406,7 @@ virCHMonitorBuildNetJson(virDomainObjPtr vm, virJSONValuePtr nets, virDomainNetD
         }
     }
     if (netdef->driver.virtio.queues) {
```
Cloud-hypervisor is checking the following:

+        if self.fds.is_some() && self.fds.as_ref().unwrap().len() * 2 != self.num_queues {
+            return Err(ValidationError::VnetQueueFdMismatch);
+        }

In the case virtio.queues is not defined in guest domain XML, I set it to default values and sent the following json to clh:

clh input json = {"cpus":{"boot_vcpus":4,"max_vcpus":4},"
console":{"mode":"Pty"},"serial":{"mode":"Null"},"memory":{"size":4294967296},"kernel":{"path":"/home/praween/libvirt/vm/hypervisor-f
w"},"disks":[{"path":"/home/praween/libvirt/vm/focal-server-cloudimg-amd64-custom.raw"},{"path":"/home/praween/libvirt/vm/cloud-confi
g-dhcp2.raw","readonly":true}],"net":[{"fds":[26],"mac":"00:11:22:39:46:11","iommu":true,"num_queues":2}]}

This handled the validation failure.

bryteise replied:
So in ch_process.c you can see where I've changed setting the tapfdSize to be = 1 where it wasn't defined (and another change to add similar logic for ethernet). Is that reasonable do you think?

praveen-pk replied:
That may be not work. Consider an configuration like below:

```
<interface type='network'>
        <mac address='00:11:22:39:46:11'/>
        <driver iommu='on'/>
      <source network='default' bridge='virbr0'/>
      <model type='virtio'/>
   </interface>
```

In this configuration cloud-hypervisor needs a TAP FD and num_queues=2 to be passed in. 

In current version, num_queues is only set when "queues" is defined in input XML.

bryteise replied:
Ahh I see. Yes num_queues isn't being passed as the check is on the netdef. I switched that over to `priv->tapfdSize` and that is happy. Unfortunately the CI isn't happy because previously cloud-hypervisor would set a default ip to the host side of the tap but now that libvirt is creating things there isn't an ip assigned so I need to make some CI updates.

praveen-pk replied:
LGTM!

bryteise committed:
Handle ethernet network interface types

These are used in the CI and were previously working when passing the
network device by tap name.

bryteise committed:
ch: Fixup formatting, correct return value

bryteise committed:
ch: Create ch num_queues based on actual fd size

The tapfdSize element of the vm's private data will always have the
correct number of queue pairs in use for the network even when queues
were not specified so use that data to set num_queues.

bryteise committed:
ch: Activate the host side of the ethernet tap device

When using an interface type='ethernet' the device can be given a host
side configuration that can be activated before starting
cloud-hypervisor (similar to how the network is activated for qemu).
